app/pages/learn/01_tutorial/04_mastering-the-api/02_invoke/00_methodhandle.md

```diff
@@ -0,0 +1,626 @@
+
+```java
+MethodHandle boundArguments = MethodHandles.insertArguments(targetMethodHandle, 1, "new", 3L);
+```
+
+To invoke the resulting adapter method handle, we only need to provide the arguments that are not pre-filled:
+
+```java
+boundArguments.invoke(1, true); // outputs: "1new3true"
+```
+
+If we try to pass the arguments that are already prefilled, we will fail with a `WrongMethodTypeException`.
+
+### Filter arguments
+We can use [`MethodHandles.filterArguments`](javadoc:MethodHandles.filterArguments(MethodHandle,int,MethodHandle...))
+to apply transformations to arguments before invocation of the target method handle. To make it work, we have to provide:
+
+- The target method handle;
+- The position of the first argument to transform;
+- Method handles for the transformations of each argument.
+
+If certain arguments don't require transformation, we can skip them by passing `null`. It's also possible to skip the
+rest of the arguments entirely if we only need to transform a subset of them.
+
+Let's reuse the method handle from the previous section and filter some its arguments before its invocation.
+
+```java
+MethodHandle targetMethodHandle = lookup.findStatic(Example.class, "test",
+        MethodType.methodType(void.class, int.class, String.class, long.class, boolean.class));
+```
+
+Let's create a method that transforms any `boolean` value by negating it:
+
+```java
+private static boolean negate(boolean original) {
+    return !original;
+}
+```
+
+and also construct a method that increments any given integer value:
+
+```java
+private static int increment(int original) {
+    return ++original;
+}
+```
+
+We can obtain method handles for these transformation methods:
+
+```java
+MethodHandle negate = lookup.findStatic(Example.class, "negate", MethodType.methodType(boolean.class, boolean.class));
+MethodHandle increment = lookup.findStatic(Example.class, "increment", MethodType.methodType(int.class, int.class));
+```
+
+and use them to get a new method handle having filtered arguments:
+
+```java
+// applies filter 'increment' to argument at index 0, 'negate' to the last argument, 
+// and passes the result to 'targetMethodHandle'
+MethodHandle withFilters = MethodHandles.filterArguments(targetMethodHandle, 0, increment, null, null, negate);
+withFilters.invoke(3, "abc", 5L, false); // outputs "4abc5true"
+```
+
+### Fold arguments
+When we want to perform pre-processing of one or more arguments before the invocation of a `MethodHandle`, we
+can use [`MethodHandles.foldArguments`](javadoc:MethodHandles.foldArguments(MethodHandle,int,MethodHandle)) and provide
+it with the method handle of any combiner method which will accept arguments starting at any preferred position.
+
+Let's assume that we have a `target` method:
+
+```java
+private static void target(int ignored, int sum, int a, int b) {
+    System.out.printf("%s + %s equals %s and %s is ignored%n", a, b, sum, ignored);
```

Comment:
While this works, it may be better to actually provide the exact format specifiers and use `%d` instead of `%s` here.